### PR TITLE
Implement `len_trim` and `trim` intrinsic

### PR DESCRIPTION
**Pull request by HarshitaKalani**
Towards: https://github.com/lfortran/lfortran/issues/3034, #4017

### CODE REVIEW

File: src/libasr/pass/intrinsic_functions.h [hunk 1]
```diff
@@ -3653,6 +3654,57 @@ namespace Adjustr {
...
+        /*
+            function len_trim(string) result(r)
+                character(len=*), intent(in) :: string
+                r = len(string)
+                if (r/= 0) then
+                    do while(string(r:r) == " ")
+                        r = r - 1
+                        if (r == 0) exit
+                    end do
+                end if
```
Comment (certik) [inline src/libasr/pass/intrinsic_functions.h:3687]: Also check if this only trims from the right, or also from the front. The implementation might need to be extended if it trims from the front too.

Comment (HarshitaKalani) [reply src/libasr/pass/intrinsic_functions.h:3687]: Yes, it returns the length of a character string, ignoring any trailing blanks.

**Comment by certik** [comment]
It seems the fn6 test is wrong, we need to initialize the variable.

**Comment by HarshitaKalani** [comment]
The following is the MRE:
```fortran
subroutine print_generic(generic)
   class(*), intent(in), optional  :: generic
   select type(generic)
      type is (character(len=*))
         print*, trim(generic)
   end select
end subroutine print_generic
```
```console
$ gfortran a.f90 && ./a.out
$ lfortran a.f90
ASR verify pass error: Unexpected args, StringTrim expects (char) as arguments
 --> a.f90:5:18
  |
5 |          print*, trim(generic)
  |                  ^^^^^^^^^^^^^ failed here


Note: Please report unclear or confusing messages as bugs at
https://github.com/lfortran/lfortran/issues.
```

**Comment by certik** [comment]
I think reference tests have to be updated.

**Comment by certik** [comment]
The issue with `fn6` is https://github.com/lfortran/lfortran/issues/3732, and the current ASR is incorrect.

Let's comment out the test, get everything passing and review. If all works, I would merge this PR.

If any CI 3rd party code breaks, then we need to fix #3732 first.

**Comment by HarshitaKalani** [comment]
The third party code breaks with the same error as mentioned here: https://github.com/lfortran/lfortran/pull/3674#issuecomment-2014339650

**Comment by certik** [comment]
Ok. Let's fix https://github.com/lfortran/lfortran/issues/3732 first then. It will not be difficult: inside the `type is (character(len=*))`, simply cast `generic` to the appropriate type everywhere it is used. Try to implement it and see.

**Comment by certik** [comment]
Thanks for implementing it, this is great (we had an incomplete implementation, now we can do it properly). There might be some issue since 3rd party codes fail. We need to figure out why and add an integration test for the failure, and fix it.

**Comment by HarshitaKalani** [comment]
Closing as implemented in https://github.com/lfortran/lfortran/pull/4938

**Pull request closed without merging**